### DISCUSSION TITLE
near-api-js / @near-js feedback research

### DISCUSSION THREAD

**Discussion by mpeterdev** [near/near-api-js#1368] [General]
The Public Goods team at Pagoda is looking to gather current impressions of `near-api-js`/`@near-js` to help inform our upcoming contributions to the interface

What have your friction points been with these libraries?
What additional functionality do you wish they offered?
What are other similar libraries you enjoy using?

Any type of input is welcome, please feel free to share rough thoughts or link to code

**Comment by bh2smith** [comment]
## Improve Return Types on Provider Methods

I find that people often write raw RPC requests instead of using this library and suspect this is because the return types of the interface are just raw json RPC responses anyway.
 
For example, gas prices: https://docs.near.org/api/rpc/gas

There appears to be this method: 

```ts
const response = await near.connection.provider.gasPrice(17824600);
```

but the return value is just:

```
{
  "jsonrpc": "2.0",
  "result": {
    "gas_price": "100000000"
  },
  "id": "dontcare"
}
```

If we are providing an API, this method should parse the response, return an actual type or error appropriately.

Not having this results in sloppy code outside of this library.

**Comment by bh2smith** [comment]
## Suggestion: Generated type interfaces for Contract instances when ABI is provided.

Currently we have to manually type our own interfaces on a per-contract basis. This is an archaic practice. Here is an example 

https://github.com/Mintbase/near-ca/blob/7546451bad872ab86329955d7f54d117f2f822f1/src/mpcContract.ts#L26-L34

**Comment by asilenced** [comment]
### Hello everyone,

I'm thrilled to participate in this discussion aimed at improving the NEAR Protocol. I've been deeply interested in NEAR Protocol and have thoroughly analyzed its core code. I recently came across the announcement about enhancing near-api-js on Telegram.

**I’d like to suggest introducing a React hooks interface.** 
Specifically, I propose adding a useNearAccount function as a React hook to manage the Account class. You can check out my implementation example here: (https://github.com/asilenced/introducing-react-hook-to-near-api-js/blob/main/account.ts).

I look forward to hearing your thoughts and feedback on this proposal.
Finally, I want to contribute to the enhancing of near-api-js.

Thank you!

**Comment by elliotBraem** [comment]
One thing that'd be great to address are the polyfills and build issues, particularly with Vite 

- [ ] @near-js+providers requires http and https node polyfills
- [ ] Warnings for js-sha256 "is strongly discouraged as it poses security risks and may cause issues with minification." (signer, transactions, wallet-selector)
- [ ] alias near-api-js in Vite app, [workaround that sometimes works](https://github.com/near/near-api-js/issues/607#issuecomment-1314301419) (although I haven't seen this issue lately, maybe it's resolved now
- [ ] have ran into buffer before, although unable to recreate now; maybe it's resolved
